feat(windows): This is to provide a PR build for testing

rc-swag opened the pull request:
These are some key tests for testing the update state machine. 
It also has a few fixes.

- Check now button uses the background update state machine instead of using the RemoteUpdateCheck unit. This now makes sure the state machine state moves to UpdateAvaiable, if there are new updates. (this should have already been checked in on the epic branch)

- The LastUpdateTimeCheck now correctly handles the case for when then the registry key does not exist. Treating it as if the time has expired since the last update.
- 
- Check for actuall class state rather than a the class name as a string.
- 
- Guard for if file has not been downloaded
- 




I am working on a spreadsheet that will be based on the state transition matrix to test more scenarios and failure states. 


# User Testing

## TEST_UI_UPDATES_KEYMAN

1. Install the Keyman build from this PR
2. Open Keyman configuration
3. In the tab check the Keyman version
4. Select the Update Tab on the left.
5. Press the `Check for updates` button
6. The screen shall be updated with the new version of Keyman Available.


## TEST_UI_UPDATES_KEYMAN_KEYMAN_HAS_RUN

1. Install the Keyman build on this PR
2. Keyman Configuration `Automatically check for updates and download` and `Start when Windows starts` is selected.
3. Start keyman 
4. We can verify the StateMachine state in the Windows Registry. 
5. <kbd>Win</kbd><kbd>R</kbd> type regedit Go the current user key `update state` 
`Computer\HKEY_CURRENT_USER\SOFTWARE\Keyman\Keyman Engine\update state`
![image](https://github.com/user-attachments/assets/7bb8023f-a409-4f66-abb3-aa4b6f12af95)
6. The state will go to `usDownloading`.
7. Press <kbd>F5</kbd>
8. When state changes to `usWaitingRestart`
9. Logout and back into Windows
10. Keyman should install.
11. Open Keyman Configuration and check that it has updated to the lastest Alpha build version.

## TEST_UI_UPDATES_KEYMAN_HAS_NOT_RUN

1. Install the Keyman build on this PR
2. Keyman Configuration `Automatically check for updates and download` and `Show splash screen` is selected. Then uncheck `Start when Windows starts`
3. Exit Keyman
4. Logout and then back into windows
5. Start keyman but at the splash screen exit
6. After a while you will get a prompt for Keyman for Windows to make changes
7. Start Keyman and check the version number is the latest Alpha build.


## TEST_UI_UPDATES_APPLY_NOW_KEYMAN

1. Follow the steps from TEST_UI_UPDATES_KEYMAN
2. Press the APPLY_UPDATE_NOW button.
3. When prompted that the computer will restart choose `update now`
4. When Finished Open Keyman Configuration and ensure the keyboard has been updated.

## TEST_UI_CHECK_KBD_UPDATES

1. Install the Keyman Attached to the PR.
2. Download the version of sil_eurolatin attached to his PR. [sil_euro_latin.zip](https://github.com/user-attachments/files/18446775/sil_euro_latin.zip)
3. Double-click on the downloaded file in Windows Explorer to install the keyboard
4. Open Keyman configuration
5. Ensure the Keyboard is installed and check the version is 3.0.0
6. Select the Update Tab on the left.
7. Press the `Check for updates` button
8. The screen shall be updated with the new version of the keyboard available.

## TEST_UI_UPDATES_APPLY_NOW_KBD

1. Follow the steps from TEST_UI_CHECK_KBD_UPDATES
2. Press the APPLY_UPDATE_NOW button.
3. When Finished Open Keyman Configuration and ensure the keyboard has been updated.

## TEST_UI_UPDATES_RESTART_KBD

1. Verify that Keyman is configured to start with Windows
2. Uninstall the sil_eurolatin Keyboard
3. Follow the steps from TEST_UI_CHECK_KBD_UPDATES
4. Logout and back into Windows
5. Open Keyman Configuration and ensure the keyboard has been updated.

ermshiperete reviewed (approved):
LGTM, although the PR title should probably be changed. It seems there are some fixes included in addition to providing a PR build for testing.

dinakaranr commented:
TEST_UI_UPDATES_KEYMAN - Passed

# Test Results
I tested this issue with the attached "18.0.169-alpha-test-12922" build(21/01/2025) on Windows 10. I'm sharing my observation here.

* **TEST_UI_UPDATES_KEYMAN (Passed):**
1. Install the Keyman attached to the "18.0.169-alpha-test-12922" PR build on  21/01/2025.
2. Open Keyman configuration
3. In the "Support" tab: Check the Keyman version
4. Select the "Update" Tab on the left.
5. Press the Check for updates button.
6. Verified that the screen was updated with the new version of Keyman.

* **TEST_UI_UPDATES_KEYMAN_KEYMAN_HAS_RUN (Passed):**
1. Install the Keyman attached to the "18.0.169-alpha-test-12922" PR build on  21/01/2025.
2. Keyman Configuration on options: "Automatically checked for updates and download" and  "Start when Windows starts" are selected.
3. Open the "Regedit" window.
4. Select the Computer\HKEY_CURRENT_USER\SOFTWARE\Keyman\Keyman Engine\update state
5. Verified that the "updatestate" is showing as "usDownloading" (It is showing a few minutes)
6. Open the keyman configuration.
7. Press the F5 in the regedit window.
8. Verified that the "updatestate" is changed to the "usWaitingRestart".
9. Logout and back into Windows
10. Keyman should install
11. Verified that the keyman was installed after clicking the "Ok" button on the "Do you want to allow this app to make changes to your device?" dialog.
12. Open Keyman Configuration.
13. Verified that the keyman has updated to the latest Alpha build version.

* **TEST_UI_UPDATES_KEYMAN_HAS_NOT_RUN (Passed):**
1. Install the Keyman attached to the "18.0.169-alpha-test-12922" PR build on  21/01/2025.
2. Keyman Configuration on options: "Automatically checked for updates and download" and  "Show splash screen" are selected. then "Start when Windows starts" is unchecked.
3. Exit the keyman.
4. Log out and then back into Windows.
5. Start keyman but at the splash screen exit
6. After a while, you will get a prompt for Keyman for Windows to make changes.
7. Verified that the "Keyman update available" dialog appeared.
8. Click the "Install" button on that dialog.
9. Start keyman and then click the "Update" tab.
10. Verified that the "Keyman" upgrade to the latest alpha build.

* **TEST_UI_UPDATES_APPLY_NOW_KEYMAN (Passed):**
1. Follow the steps from the TEST_UI_UPDATES_KEYMAN scenario.
2. Press the APPLY_UPDATE_NOW button.
3. Verified that the "Keyman update" dialog appeared
4. When prompted the computer will restart. I choose "Update now"
5. "Do you want to allow this app to make changes to your device?" dialog appeared
6. Press the "Yes" button on that dialog.
7. The "Keyman Setup" dialog appeared
8. Press the "Yes" button.
9. The system is restarted and then open the "Keyman configuration" dialog.
10. Verified that the "Keyman" upgrade to the latest alpha build.

* **TEST_UI_CHECK_KBD_UPDATES (Passed):**
1. Install the Keyman attached to the "18.0.169-alpha-test-12922" PR build on  21/01/2025.
2. Download the version of sil_eurolatin attached to his PR. sil_euro_latin.zip
3. Double-click on the downloaded file in Windows Explorer to install the keyboard
4. Open Keyman configuration
5. Ensure the Keyboard is installed and check the version is 3.0.0
6. Select the Update Tab on the left.
7. Press the "Check for updates" button
8. Verified that the "EuroLatin" keyboard is updated version as 3.0.3 in the table. 

* **TEST_UI_UPDATES_APPLY_NOW_KBD (Passed):**
1. Follow the steps from TEST_UI_CHECK_KBD_UPDATES
2. Press the APPLY_UPDATE_NOW button.
3. Verified that the "Keyman update" dialog appeared.
4. Press the "Update now" on that dialog.
5. "Do you want to allow this app to make changes to your device?" dialog appeared
6. Press the "Yes" button on that dialog.
7. Verified that the "Eurolatin" keyboard upgraded from 3.0.0 to 3.0.3 version.

* **TEST_UI_UPDATES_RESTART_KBD (Passed):**
1. Keyman Configuration on options: "Automatically checked for updates and download" and  "Show splash screen" is selected. then "Start when Windows starts" is checked.
2. Uninstall the sil_eurolatin Keyboard
3. Follow the steps from TEST_UI_CHECK_KBD_UPDATES
4. The system is restarted and then open the "Keyman configuration" dialog.
5. Verified that the "Eurolatin" keyboard upgraded from 3.0.0 to 3.0.3 version.

rc-swag commented:
> the PR title should probably be changed. It seems there are some fixes included in addition ..

Yes thanks for pointing that out. I updated the title but also updated the PR comment block to include a summary of the additional fixes.
